app/policies/api/rubygem_policy.rb

```diff
@@ -5,6 +5,19 @@ class Scope < Api::ApplicationPolicy::Scope
   alias rubygem record
 
   def show_trusted_publishers?
-    api_key_scope?(:configure_trusted_publishers, rubygem) && user_policy!.show_trusted_publishers?
+    api_key_scope?(:configure_trusted_publishers, rubygem) &&
+      user_authorized?(rubygem, :show_trusted_publishers?)
+  end
+
+  def add_owner?
+    mfa_requirement_satisfied?(rubygem) &&
```

Comment:
before wasn't the mfa_requirement the last thing checked rather than the first?

Reply:
The original order is as follows (with solution to the failure code in comment):

```ruby
authenticate_with_api_key # fail reveals public info (api key is invalid)
find_rubygem              # fail reveals public info (gem is not found)
verify_user_api_key       # fail reveals      ? info (type of api key)
verify_mfa_requirement    # fail reveals public info (gem download numbers)
verify_api_key_gem_scope  # fail reveals      ? info (scopes in the api key)
verify_gem_ownership      # fail reveals public info (gem ownership)
verify_with_otp           # fail reveals public info (otp is invalid)
                          # success indicates full authentication
```

I think, given the above the correct order should be (and I will update the PR to):

```ruby
authenticate_with_api_key # [user] make a new api key
find_rubygem              # [gem] use the correct url
verify_with_otp           # [user] submit the correct otp (bypasses if mfa is not enabled)
verify_mfa_requirement    # [gem + user] set up MFA
verify_user_api_key       # [apikey] use a different process (non-automated)
verify_api_key_gem_scope  # [gem + apikey] create a new api key with the correct scope
verify_gem_ownership      # [gem + user] ask for ownership of the gem (we should not show this until we have fully authenticated)
```

Given this proposed order, we would not reveal anything about the access a key might provide without the user submitting otp correctly when enabled. For gems requiring mfa, we would not reveal whether the api key had access to the gem before the user enabled mfa. Submitting OTP and enabling MFA are authn steps that should happen before we reveal authz of the particular key in use, imho.

However, this is all speculation and I could be misguided here. Overall, I think we should tell them that they won't be able to access this gem without account level changes (mfa, ownership) before we tell them that this api key is misconfigured in some way.